Active-Active Region Application Failover (#1034)
RedisLabs/redislabs-docs: content/rs/references/developing-for-active-active/region-application-failover-active-active.md

1. **Local failures** - The local replica is down or otherwise unavailable
1. **Replication failures** - The local replica is available but fails to replicate writes to or from remote replicas

### Local Failures

Local failure detection is when the application is unable to connect to the Redis endpoint for any reason, for example: multiple node failures, configuration errors, connection refused, connection timed out, unexpected protocol level errors.

### Replication Failures

Replication failures are more difficult to detect reliably without causing false positives. Replication failures can include: network split, replication configuration issues, remote replica failures.

The most reliable method for health-checking replication is by using the Redis publish/subscribe (pub/sub) mechanism.

{{< note >}}
Note that this document does not suggest that Redis pub/sub is reliable in the common sense. Messages can get lost in certain conditions, but that is acceptable in this case because typically the application determines that replication is down only after not being able to deliver a number of messages over a period of time.
{{< /note >}}

> When you use the pub/sub data type to detect failures:

Reviewer: ```suggestion
When you use the pub/sub data type to detect failures, the application:
```
```suggestion
When you use the pub/sub data type to detect failures:
```

> 1. The application establishes a connection to all replicas and subscribes to a dedicated per-replica channel.

Reviewer: ```suggestion
1. Connects to all replicas and subscribes to a dedicated channel for each replica.
```

> 1. The application establishes another connection to all replicas and periodically publishes a uniquely identifiable message.

Reviewer: ```suggestion
1. Connects to all replicas and periodically publishes a uniquely identifiable message.
```
```suggestion
1. The application establishes another connection to all replicas and periodically publishes a uniquely identifiable message.
```

> 1. The application monitors received messages and ensures that it is able to receive its own messages within a predetermined window of time.

Reviewer: ```suggestion
1. Monitors received messages and ensures that it is able to receive its own messages within a predetermined window of time.
```
```suggestion
1. The application monitors received messages and ensures that it is able to receive its own messages within a predetermined window of time.
```

> You can also rely on known dataset changes to monitor the reliability of the replication stream,

Reviewer: ```suggestion
You can also use known dataset changes to monitor the reliability of the replication stream,
```
```suggestion
You can also rely on known dataset changes to monitor the reliability of the replication stream,
```
